### PR TITLE
feat: start region alive keepers

### LINKED ISSUE #1737: Region failover causes region is opened on multiple datanodes
### What type of bug is this?

Data corruption

### What subsystems are affected?

Datanode

### What happened?

Region `6103148527616` (table `cbmx4k0furmt.dbtest.node_xfs_allocation_btree_records_deleted_total`) kept migrating between datanodes and at some time it got opened on both datanode-0 and datanode-2 simutaneously. These two datanodes both carried out compaction, causing the sequence became discontinous.

Compaction on datanode-2
image

At the same time, compaction on datanode-0
image

Then the memtable panics:
image

### What operating system did you use?

NA

### Relevant log output and stack trace

```bash
[datanode-2] 2023-06-06 16:18:22	
2023-06-06T08:18:22.370746Z  INFO storage::compaction::scheduler: Successfully compacted region: 6103148527616

[datanode-0] 2023-06-06 16:18:23	
2023-06-06T08:18:23.769238Z  INFO storage::compaction::scheduler: Successfully compacted region: 6103148527616
```

### How can we reproduce the bug?

I have no idea on how to reproduce this bug.

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

fix #1737 

- start region alive keepers when table registered or region opened
- add more tests

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/catalog/src/remote/region_alive_keeper.rs [hunk 1]
```diff
@@ -114,10 +129,11 @@ impl RegionAliveKeepers {
...
     }
 
-    pub async fn start(&self, heartbeat_interval_millis: u64) {
+    pub async fn start(&self) {
         for keeper in self.keepers.lock().await.values() {
-            keeper.start(heartbeat_interval_millis).await;
+            keeper.start().await;
         }
+        self.started.store(true, Ordering::Relaxed);
     }
```
Comment (WenyXu) [inline src/catalog/src/remote/region_alive_keeper.rs:137]: I guess we should execute `self.started.store(true, Ordering::Relaxed);` first?
```suggestion
    pub async fn start(&self) {
        self.started.store(true, Ordering::Relaxed);
        for keeper in self.keepers.lock().await.values() {
            keeper.start().await;
        }
    }
```

File: src/catalog/src/remote/region_alive_keeper.rs [hunk 2]
```diff
@@ -235,10 +263,12 @@ impl RegionAliveKeeper {
...
 
-    async fn start(&self, heartbeat_interval_millis: u64) {
+    async fn start(&self) {
         for handle in self.countdown_task_handles.lock().await.values() {
-            handle.start(heartbeat_interval_millis).await;
+            handle.start(self.heartbeat_interval_millis).await;
         }
+
+        self.started.store(true, Ordering::Relaxed);
```
Comment (WenyXu) [inline src/catalog/src/remote/region_alive_keeper.rs:271]: ditto

File: src/catalog/src/remote/region_alive_keeper.rs (hunk 1 above)
Comment (MichaelScofield) [reply src/catalog/src/remote/region_alive_keeper.rs:137]: The region alive keeper will be first inserted into `keepers`, then check start. See `register_table` method. If we set `started = true` first here, the keeper might be started twice.

File: src/catalog/src/remote/region_alive_keeper.rs (hunk 2 above)
Comment (MichaelScofield) [reply src/catalog/src/remote/region_alive_keeper.rs:271]: same as above

File: src/catalog/src/remote/region_alive_keeper.rs (hunk 1 above)
Comment (WenyXu) [reply src/catalog/src/remote/region_alive_keeper.rs:137]: Is it possible that set `stated` later might cause some keepers don't start?

File: src/catalog/src/remote/region_alive_keeper.rs [hunk 3]
```diff
@@ -213,11 +237,15 @@ impl RegionAliveKeeper {
...
-            .insert(region, handle);
+            .insert(region, handle.clone());
         info!(
             "Register alive countdown for new region {region} in table {}",
             self.table_ident
-        )
+        );
+
+        if self.started.load(Ordering::Relaxed) {
+            handle.start(self.heartbeat_interval_millis).await;
```
Comment (fengjiachun) [inline src/catalog/src/remote/region_alive_keeper.rs:247]: If there is no `start`, should print some logs.

File: src/catalog/src/remote/region_alive_keeper.rs [hunk 4]
```diff
@@ -78,17 +82,28 @@ impl RegionAliveKeepers {
...
         }
 
         info!("Register RegionAliveKeeper for table {table_ident}");
-        self.keepers.lock().await.insert(table_ident, keeper);
+        self.keepers
+            .lock()
+            .await
+            .insert(table_ident, keeper.clone());
+
+        if self.started.load(Ordering::Relaxed) {
```
Comment (fengjiachun) [inline src/catalog/src/remote/region_alive_keeper.rs:100]: If there is no `start`, should print some logs.

**Review by fengjiachun — APPROVED** [review]

**Review by WenyXu — APPROVED** [review]

**Pull request merged**